trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/ACSLHandler.java

```diff
@@ -212,32 +222,80 @@ public Result visit(final IDispatcher main, final ACSLNode node) {
-			for (final Overapprox overapprItem : resultBuilder.getOverappr()) {
-				overapprItem.annotate(assertStmt);
+			final CodeStatement codeStmt = ((CodeAnnotStmt) node).getCodeStmt();
+			if (codeStmt instanceof Assertion) {
+				ExpressionResult formula =
+						(ExpressionResult) main.dispatch(((Assertion) codeStmt).getFormula(), main.getAcslHook());
+
+				formula = mExprResultTransformer.transformSwitchRexIntToBool(formula, loc, main.getAcslHook());
+
+				resultBuilder.addAllExceptLrValue(formula);
+
+				final AssertStatement assertStmt = new AssertStatement(loc, formula.getLrValue().getValue());
+				// TODO: Handle havoc statements
+				for (final Overapprox overapprItem : resultBuilder.getOverappr()) {
+					overapprItem.annotate(assertStmt);
+				}
+				resultBuilder.addStatement(assertStmt);
+				final List<HavocStatement> havocs = CTranslationUtil.createHavocsForAuxVars(formula.getAuxVars());
+				resultBuilder.addStatements(havocs);
+				final Check check;
+				if (mWitnessInvariantMode) {
+					check = new Check(Spec.WITNESS_INVARIANT);
+				} else {
+					check = new Check(Spec.ASSERT);
+				}
+				check.annotate(assertStmt);
+			}
+			if (codeStmt instanceof GhostUpdate) {
+				final GhostUpdate update = (GhostUpdate) codeStmt;
+				final ExpressionResult exprResult =
+						(ExpressionResult) main.dispatch(update.getExpr(), main.getAcslHook());
+				resultBuilder.addAllExceptLrValue(exprResult);
+				final SymbolTableValue stv = mSymboltable.findCSymbol(main.getAcslHook(), update.getIdentifier());
+				if (stv == null) {
+					throw new IncorrectSyntaxException(loc,
+							"Undeclared variable in ACSL expression: " + update.getIdentifier());
+				}
+				if (!stv.getBoogieName().startsWith(SFO.GHOST)) {
+					throw new IncorrectSyntaxException(loc,
+							"C variable " + update.getIdentifier() + " cannot be assigned in ghost statement.");
+				}
+				final VariableLHS lhs = new VariableLHS(loc, mTypeHandler.getBoogieTypeForCType(stv.getCType()),
+						stv.getBoogieName(), stv.getDeclarationInformation());
+				resultBuilder.addStatement(StatementFactory.constructSingleAssignmentStatement(loc, lhs,
+						exprResult.getLrValue().getValue()));
+			}
+			if (codeStmt instanceof GhostDeclaration) {
+				final GhostDeclaration decl = (GhostDeclaration) codeStmt;
+				final String boogieName = SFO.GHOST + decl.getIdentifier();
+				final CPrimitive cType = AcslTypeUtils.translateAcslTypeToCType(decl.getType());
+				final ASTType astType = mTypeHandler.cType2AstType(loc, cType);
+				final Declaration boogieDecl = new VariableDeclaration(loc, new Attribute[0],
+						new VarList[] { new VarList(loc, new String[] { boogieName }, astType) });
+				final CDeclaration cDecl = new CDeclaration(cType, decl.getIdentifier());
+				final IASTFunctionDefinition scope = CdtASTUtils.findScope(main.getAcslHook());
+				DeclarationInformation declInfo;
+				if (scope == null) {
+					declInfo = DeclarationInformation.DECLARATIONINFO_GLOBAL;

```

Comment:
Can this happen even if a declaration is not parsed as `GlobalGhostDeclaration`?

Reply:
No, this should not happen. However, it might happen if we instrument witnesses as ghost-variables, because everything there is handled by `ACSLHandler`.